src/actions.ts

```diff
@@ -0,0 +1,449 @@
+			(params.p1.length > 18) ||
+			(params.p2.length > 18) ||
+			// the battle ID must be valid
+			!extractedFormatId ||
+			// the format ID must be valid
+			!formatId ||
+			// the format from the battle ID must match the format ID
+			(formatId[1] !== extractedFormatId[1])
+		) {
+			return 0;
+		}
+
+		if (server.id !== Config.mainserver) {
+			params.id = server.id + '-' + params.id;
+		}
+		params.serverid = server.id;
+
+		const result = await Replays.prep(params);
+
+		this.setPrefix(''); // No need for prefix since only usable by server.
+		return result;
+	},
+	uploadreplay(params) {
+		this.setHeader('Content-Type', 'text/plain; charset=utf-8');
+		return Replays.upload(params, this);
+	},
+	async invalidatecss() {
+		const server = await this.getServer(true);
+		if (!server) {
+			return {errorip: this.getIp()};
+		}
+		// No need to sanitise server['id'] because it should be safe already.
+		const cssfile = `${__dirname}/../config/customcss/${server['id']}.css`;
+		return new Promise<{actionsuccess: boolean}>(resolve => {
+			fs.unlink(cssfile, err => resolve({actionsuccess: !err}));
+		});
+	},
+	async changepassword(params) {
+		if (this.request.method !== 'POST') {
+			throw new ActionError(`'changepassword' requests can only be made with POST data.`);
+		}
+		if (!params.oldpassword) {
+			throw new ActionError(`Specify your current password.`);
+		}
+		if (!params.password) {
+			throw new ActionError(`Specify your new password.`);
+		}
+		if (!params.cpassword) {
+			throw new ActionError(`Repeat your new password.`);
+		}
+
+		if (!this.user.loggedin) {
+			throw new ActionError('Your session has expired. Please log in again.');
+		}
+		if (params.password !== params.cpassword) {
+			throw new ActionError('Your new passwords do not match.');
+		}
+		if (!(await this.session.passwordVerify(this.user.id, params.oldpassword))) {
+			throw new ActionError('Your old password was incorrect.');
+		}
+		params.password = params.password.replace(/\s/ig, '');
+		if (params.password.length < 5) {
+			throw new ActionError('Your new password must be at least 5 characters long.');
+		}
+		const actionsuccess = await this.session.changePassword(this.user.id, params.password);
+		return {actionsuccess};
+	},
+	async changeusername(params) {
+		if (this.request.method !== 'POST') {
+			throw new ActionError('Invalid request (username changing must be done through POST requests).');
+		}
+		if (!params.username) {
+			throw new ActionError(`Specify a username.`);
+		}
+		if (!this.user.loggedin) {
+			throw new ActionError('Your session has expired. Please log in again.');
+		}
+		if (toID(params.username) !== this.user.id) {
+			throw new ActionError('You\'re not logged in as that user.');
+		}
+		// safe to use userid directly because we've confirmed they've logged in.
+		const actionsuccess = await tables.users.update(this.user.id, {
+			username: params.username,
+		}).catch(() => false);
```

Comment:
Why silence exceptions? It sounds like the return type of this function is not adequate